contracts/utils/math/Math.sol

```diff
@@ -339,38 +339,77 @@ library Math {
-        // `msb(a) <= a < 2*msb(a)`. This value can be written `msb(a)=2**k` with `k=log2(a)`.
-        //
-        // This can be rewritten `2**log2(a) <= a < 2**(log2(a) + 1)`
-        // → `sqrt(2**k) <= sqrt(a) < sqrt(2**(k+1))`
-        // → `2**(k/2) <= sqrt(a) < 2**((k+1)/2) <= 2**(k/2 + 1)`
-        //
-        // Consequently, `2**(log2(a) / 2)` is a good first approximation of `sqrt(a)` with at least 1 correct bit.
-        uint256 result = 1 << (log2(a) >> 1);
-
-        // At this point `result` is an estimation with one bit of precision. We know the true value is a uint128,
-        // since it is the square root of a uint256. Newton's method converges quadratically (precision doubles at
-        // every iteration). We thus need at most 7 iteration to turn our partial result with one bit of precision
-        // into the expected uint128 result.
         unchecked {
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            return min(result, a / result);
+            // Take care of easy edge cases when a == 0 or a == 1
+            if (a <= 1) {
+                return a;
+            }
+            // This check ensures no overflow at return; see below.
+            if (a >= uint256(type(uint128).max) ** 2) {
+                return type(uint128).max;
+            }
+
+            uint256 aAux = a;
+            uint256 result = 1; // Initial approximate bit length.
+
+            // For the first guess of `result` (e), we get the biggest power of 2 which is smaller than sqrt(a)
+            // (i.e. 2^e <= sqrt(a)). We know that e is at most 127 given (2^128)^2 overflows an uint256.
+            // Thus, we approximate e by iterating 2^{i/2} where i starts at 128, and applying the exponent
+            // to e if the result is still smaller than a (up to e == 127).
+            if (aAux >= (1 << 128)) {
+                aAux >>= 128;
+                result = 1 << 64;
+            }
+            if (aAux >= (1 << 64)) {
+                aAux >>= 64;
+                result <<= 32;
+            }
+            if (aAux >= (1 << 32)) {
+                aAux >>= 32;
+                result <<= 16;
+            }
+            if (aAux >= (1 << 16)) {
+                aAux >>= 16;
+                result <<= 8;
+            }
+            if (aAux >= (1 << 8)) {
+                aAux >>= 8;
+                result <<= 4;
+            }
+            if (aAux >= (1 << 4)) {
+                aAux >>= 4;
+                result <<= 2;
+            }
+            if (aAux >= (1 << 2)) {
+                result <<= 1;
+            }
```

Comment:
Here we should use the branchless approach from the log functions

```suggestion
            uint256 isGt;

            uint256 isGt = SafeCast.toUint(aAux > (1 << 128) - 1);
            aAux >>= isGt * 128;
            result <<= isGt * 64;

            uint256 isGt = SafeCast.toUint(aAux > (1 << 64) - 1);
            aAux >>= isGt * 64;
            result <<= isGt * 32;

            uint256 isGt = SafeCast.toUint(aAux > (1 << 32) - 1);
            aAux >>= isGt * 32;
            result <<= isGt * 16;

            uint256 isGt = SafeCast.toUint(aAux > (1 << 16) - 1);
            aAux >>= isGt * 16;
            result <<= isGt * 8;

            uint256 isGt = SafeCast.toUint(aAux > (1 << 8) - 1);
            aAux >>= isGt * 8;
            result <<= isGt * 4;

            uint256 isGt = SafeCast.toUint(aAux > (1 << 4) - 1);
            aAux >>= isGt * 4;
            result <<= isGt * 2;

            uint256 isGt = SafeCast.toUint(aAux > (1 << 2) - 1);
            result <<= isGt * 1;
```

Reply:
I'm also getting that the current implementation is better:

```
FOUNDRY_FUZZ_RUNS=10000 forge snapshot --via-ir --optimizer-runs 1000000000 --use 0.8.24 --match-test testSqrt -vvvv

## Current
MathTest:testSqrt(uint256,uint8) (runs: 10000, μ: 4823, ~: 4986)

## New
MathTest:testSqrt(uint256,uint8) (runs: 10000, μ: 4883, ~: 5012)
```

```
FOUNDRY_FUZZ_RUNS=10000 forge snapshot --use 0.8.24 --match-test testSqrt -vvvv

## Current
MathTest:testSqrt(uint256,uint8) (runs: 10000, μ: 5143, ~: 5302)

## New
MathTest:testSqrt(uint256,uint8) (runs: 10000, μ: 5225, ~: 5455)
```